### ISSUE TITLE
Enforce SMTP transport encryption with DNS-based Authentication of Named Entities

### ISSUE THREAD

**Issue by renne** [YunoHost/issues#2303]
Yunohost uses StartTLS to hand-shake TLS encryptions, which is prone to downgrade attacks.
[DANE-SMTP](https://en.wikipedia.org/wiki/DNS-based_Authentication_of_Named_Entities) was developed to prevent these downgrade attacks.
In DANE-SMTP the SMTP server announces the DANE requirement via a TLSA record in the DNS zone.
If a SMTP client supports DANE-SMTP it will only connect if the SSL certificate of the SMTP server matches the TLSA record and TLS encryption is enforced.

Please enable DANE-SMTP in Yunohost.

**Comment by renne** [comment]
Outgoing SMTP with DANE:
Add file `/etc/dnsmasq.d/dnssec` to enable [DNSSEC](https://de.wikipedia.org/wiki/Domain_Name_System_Security_Extensions) validation:
```
conf-file=/usr/share/dnsmasq-base/trust-anchors.conf
dnssec
```
Changes in `/etc/postfix/main.cf`:
```
smtp_dns_support_level = dnssec
smtp_tls_security_level = dane
```

**Comment by ericgaspar** [comment]
why ping @tituspijean?

**Comment by renne** [comment]
Dumb idea ;-)

**Comment by renne** [comment]
Incoming SMTP with DANE:
1. Enable DNSSEC at your domain registrar.
1. Install debian package `ldnsutils`
1. When the certificate changes replace the TLSA record for the DNS zone: `ldns-dane create -c /etc/yunohost/certs/<domain>/crt.pem <domain> 443`
1. Create a CNAME record for each open SSL-Port following the scheme `_<port>._<tcp|udp>  3600  IN  CNAME  _443._tcp.<domain>.`

**Comment by tituspijean** [comment]
> Dumb idea ;-)

Most likely, though I am curious about your thought process.

> It is required in Germany

No. These are guidelines. https://www.bsi.bund.de/EN/Themen/Unternehmen-und-Organisationen/Standards-und-Zertifizierung/Technische-Richtlinien/technische-richtlinien_node.html

> As it is state of the art it is also required by the [General Data Protection Regulation](https://gdpr.eu/) of the European Union.

Is there a reliable source about that, or are you namedropping the GDPR here?

---

Please stay technical and do not try to "impress" contributors. It does not work.
If I am wrong, do correct me. If I am right, please edit your first message and we will cleanup this thread.

**Comment by renne** [comment]
> > Dumb idea ;-)
> 
> Most likely, though I am curious about your thought process.

Probably just the habit of addressing people in forums. And this is not a forum
 
> > It is required in Germany
> 
> No. These are guidelines. https://www.bsi.bund.de/EN/Themen/Unternehmen-und-Organisationen/Standards-und-Zertifizierung/Technische-Richtlinien/technische-richtlinien_node.html
> 
> > As it is state of the art it is also required by the [General Data Protection Regulation](https://gdpr.eu/) of the European Union.
> 
> Is there a reliable source about that, or are you namedropping the GDPR here?
> 
> Please stay technical and do not try to "impress" contributors. It does not work. If I am wrong, do correct me. If I am right, please edit your first message and we will cleanup this thread.

To avoid a lengthy legal discussion about articles 25 and 32 of the GPDR I have removed the legal reason for this issue.

**Comment by TabTwo** [comment]
One constraint that I think decides the design here, and which isn't mentioned yet in the thread: **YunoHost generates a fresh private key on every certificate renewal.**

In `src/certificate.py`, `_fetch_and_enable_new_certificate()` calls `_generate_key(domain_key_file)` unconditionally before building the CSR — there is no equivalent of certbot's `--reuse-key`. So the leaf public key changes every ~60–90 days.

That matters because the obvious TLSA record to publish is `3 1 1` (DANE-EE, SPKI, SHA-256), which pins exactly that public key. Published naively, it becomes wrong at every renewal, and the failure mode is bad: DANE-validating senders stop delivering, silently, with no visible change on our side. `@renne`'s step "when the certificate changes replace the TLSA record" is correct, but as a manual step on a self-hosted box it is an outage generator rather than a feature.

So I think incoming DANE reduces to a key-lifecycle decision, and there are three options:

1. **Reuse the key across renewals.** Smallest change, makes `3 1 1` stable. Costs automatic key rotation, which could be reintroduced deliberately with a rollover rather than happening implicitly every renewal.
2. **Publish `2 1 1` (DANE-TA)** pinning the CA instead of the leaf. Survives key rotation, but breaks when Let's Encrypt rotates intermediates — which they do, and which we do not control.
3. **Proper rollover:** generate the next key ahead of time and publish TLSA records for both current and next during the transition. Correct, and the most work — it needs the next key to exist before renewal and DNS updates to be reliable.

Whichever is chosen, TLSA publication has to be automated alongside the existing DNS push, or it will be a footgun. Worth flagging that this path already has rough edges: the lexicon-based push currently mangles MX and SRV records for at least one registrar (I filed #2830, and dns-lexicon#102 for the underlying priority-field bug), so "YunoHost writes the record for you" is not yet something to lean on for a record whose incorrectness silently stops mail.

**Suggestion: split this issue.** The outgoing half is genuinely easy and independent of all of the above —

```
smtp_dns_support_level = dnssec
smtp_tls_security_level = dane
```

plus a validating resolver. It only requires us to *validate other people's* TLSA records, publishes nothing, and cannot break our own mail delivery. It could ship well before the incoming half, which is blocked on the key-lifecycle question.

Happy to test: I run a mail server with DNSSEC on the domains and have been working through DANE/TLSA there, so I can try whichever approach lands and report on real delivery.

**Comment by Josue-T** [comment]
> Suggestion: split this issue. The outgoing half is genuinely easy and independent of all of the above —

Done with #2842. Following this, the idea is that this issue is dedicated to the client side configuration.

**Comment by Josue-T** [comment]
Soo, about 

```
smtp_dns_support_level = dnssec
smtp_tls_security_level = dane
```

My opinion about this is that this option will **not** improve mail delivery. Even worse, this option might create truble on delivery, because if the target server has a wrong TLSA configuration the email will never arrive. So this is a nice option but it must be optional. Some year ago I did some test around this, I enabled theses option:
```
smtp_dns_support_level=dnssec 
smtp_host_lookup=dns 
smtp_tls_security_level = dane 
smtp_tls_loglevel=1 
```
but at the end I removed because it was creating truble on mail delivery because many sysadmin don't really take care of having this correctly configured.  For instance, I discovered that the domain login.infomaniak.com has Dane broken (for https), but it's still be an example and probably many other domain are broken like this for a while.

So in conclusion, Dane/TLSA is great, but in really it's difficult to enforce it because it might make a lot of issue. But if you want to enforce security and you are ok to lose some email in this case it would be a good choice to enable it.
